File: src/app.test.ts
import { afterEach, beforeEach, describe, it } from "@std/testing/bdd";
import { assertSpyCall } from "@std/testing/mock";
import {
  assertSpyCommandsCalls,
  assertSpyDisplaysCalls,
  assertSpyLoggerCalls,
  makeSpyCommands,
  makeSpyDisplays,
  makeSpyLogger,
  type SpyCommands,
  type SpyDisplays,
  type SpyLogger,
} from "./core/test-utils.ts";
import { main } from "./app.ts";
import { deepMerge } from "@std/collections";

type DeepPartial<T> = {
  [K in keyof T]?: T[K] extends object ? DeepPartial<T[K]> : T[K];
};

let workingDir: string;
let logger: SpyLogger;
let displays: SpyDisplays;
let commands: SpyCommands;

beforeEach(async () => {
  workingDir = await Deno.makeTempDir();
  logger = makeSpyLogger();
  displays = makeSpyDisplays();
  commands = makeSpyCommands();
});

afterEach(async () => {
  await Deno.remove(workingDir, { recursive: true });
});

const makeMainArgs = (
  args?: DeepPartial<Parameters<typeof main>[0]>,
) =>
  (deepMerge({
    dotfilesFilesDir: workingDir,
    homedir: workingDir,
    logger,
    commands,
    displays,
    cmdArgs: {
      rest: [],
      flags: {
        printEnvironment: false,
        color: false,
        help: false,
        version: false,
      },
    },
  }, args ?? {})) as Parameters<typeof main>[0];

describe("main()", () => {
  it("should print help if `help` flag is provided", async () => {
    await main(
      makeMainArgs({ cmdArgs: { flags: { help: true } } }),
    );

    assertSpyDisplaysCalls(displays, { help: 1 });
    assertSpyCommandsCalls(commands);
  });

  it("should print environment if `printEnvironment` flag is provided", async () => {
    await main(
      makeMainArgs({ cmdArgs: { flags: { printEnvironment: true } } }),
    );

    assertSpyDisplaysCalls(displays, { environment: 1 });
    assertSpyCommandsCalls(commands);
  });

  it("should print version if `version` flag is provided", async () => {
    await main(
      makeMainArgs({ cmdArgs: { flags: { version: true } } }),
    );

    assertSpyDisplaysCalls(displays, { version: 1 });
    assertSpyCommandsCalls(commands);
  });

  it("should run diff if `diff` command provided", async () => {
    await main(
      makeMainArgs({
        homedir: "foo",
        dotfilesFilesDir: "bar",
        cmdArgs: { rest: ["diff"] },
      }),
    );

    assertSpyDisplaysCalls(displays);
    assertSpyCommandsCalls(commands, { diff: 1 });
    assertSpyCall(commands.diff, 0, {
      args: [{
        fromDir: "bar",
        toDir: "foo",
        extra: {
          diff: { color: false },
          logger,
          color: false,
        },
      }],
    });
  });

  it("should run apply if `apply` command provided", async () => {
    await main(
      makeMainArgs({ cmdArgs: { rest: ["apply"] } }),
    );

    assertSpyDisplaysCalls(displays);
    assertSpyCommandsCalls(commands, { apply: 1 });
    assertSpyCall(commands.apply, 0, {
      args: [{
        from: undefined,
        extra: {
          homedir: workingDir,
          dotfilesFilesDir: workingDir,
          logger,
          color: false,
        },
      }],
    });
  });

  it("should run apply with first arg if `apply` command provided with an arg", async () => {
    await main(
      makeMainArgs({ cmdArgs: { rest: ["apply", "foo"] } }),
    );

    assertSpyDisplaysCalls(displays);
    assertSpyCommandsCalls(commands, { apply: 1 });
    assertSpyCall(commands.apply, 0, {
      args: [{
        from: "foo",
        extra: {
          homedir: workingDir,
          dotfilesFilesDir: workingDir,
          logger,
          color: false,
        },
      }],
    });
  });

  it("should run adopt if `adopt` command provided", async () => {
    await main(
      makeMainArgs({ cmdArgs: { rest: ["adopt"] } }),
    );

    assertSpyDisplaysCalls(displays);
    assertSpyCommandsCalls(commands, { adopt: 1 });
    assertSpyCall(commands.adopt, 0, {
      args: [{
        from: undefined,
        extra: {
          homedir: workingDir,
          dotfilesFilesDir: workingDir,
          logger,
          color: false,
        },
      }],
    });
  });

  it("should run adopt with first arg if `adopt` command provided with an arg", async () => {
    await main(
      makeMainArgs({ cmdArgs: { rest: ["adopt", "foo"] } }),
    );

    assertSpyDisplaysCalls(displays);
    assertSpyCommandsCalls(commands, { adopt: 1 });
    assertSpyCall(commands.adopt, 0, {
      args: [{
        from: "foo",
        extra: {
          homedir: workingDir,
          dotfilesFilesDir: workingDir,
          logger,
          color: false,
        },
      }],
    });
  });

  it("should print help if command is not supported", async () => {
    await main(
      makeMainArgs({ cmdArgs: { rest: ["foo"] } }),
    );

    assertSpyCommandsCalls(commands);
    assertSpyDisplaysCalls(displays, { help: 1 });
    assertSpyLoggerCalls(logger, { log: { warn: 1 } });
    assertSpyCall(logger.log.warn, 0, { args: ["Command foo not found"] });
  });
});

File: src/app_test.go
package src_test

import (
	"github.com/m4rc3l05/dots/src"
	"github.com/m4rc3l05/dots/src/commands"
	"github.com/m4rc3l05/dots/src/testing"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

var _ = Describe("app()", func() {
	var cmds *testing.SpyCommands
	var displays *testing.SpyDisplays
	var logger *testing.SpyLogger

	_ = BeforeEach(func() {
		logger = testing.MakeSpyLogger()
		cmds = testing.MakeSpyCommands()
		displays = testing.MakeSpyDisplays()
	})

	It("should print help if `help` flag is provided", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Flags: src.CmdFlagsArgs{
					Help: true,
				},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, nil)
		testing.AssertSpyCommandsCalls(*cmds, nil)
		testing.AssertSpyDisplaysCalls(*displays, &testing.SpyDisplaysCallNumber{Help: 1})
	})

	It("should print environment if `printEnvironment` flag is provided", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Flags: src.CmdFlagsArgs{
					PrintEnvironment: true,
				},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, nil)
		testing.AssertSpyCommandsCalls(*cmds, nil)
		testing.AssertSpyDisplaysCalls(*displays, &testing.SpyDisplaysCallNumber{Environment: 1})
	})

	It("should print version if `version` flag is provided", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Flags: src.CmdFlagsArgs{
					Version: true,
				},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, nil)
		testing.AssertSpyCommandsCalls(*cmds, nil)
		testing.AssertSpyDisplaysCalls(*displays, &testing.SpyDisplaysCallNumber{Version: 1})
	})

	It("should run diff if `diff` command provided", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Rest: []string{"diff"},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, nil)
		testing.AssertSpyCommandsCalls(*cmds, &testing.SpyCommandsCallNumber{Diff: 1})
		testing.AssertSpyDisplaysCalls(*displays, nil)
	})

	It("should run apply if `apply` command provided", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Rest: []string{"apply"},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, nil)
		testing.AssertSpyCommandsCalls(*cmds, &testing.SpyCommandsCallNumber{Apply: 1})
		testing.AssertSpyDisplaysCalls(*displays, nil)
	})

	It("should run apply with first arg if `apply` command provided with an arg", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Rest: []string{"apply", "foo"},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, nil)
		testing.AssertSpyCommandsCalls(*cmds, &testing.SpyCommandsCallNumber{Apply: 1})
		testing.AssertSpyDisplaysCalls(*displays, nil)
		Expect(cmds.Calls.Apply[0].Args).To(Equal([]any{commands.ApplyArgs{From: "foo"}}))
	})

	It("should run adopt if `adopt` command provided", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Rest: []string{"adopt"},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, nil)
		testing.AssertSpyCommandsCalls(*cmds, &testing.SpyCommandsCallNumber{Adopt: 1})
		testing.AssertSpyDisplaysCalls(*displays, nil)
	})

	It("should run adopt with first arg if `adopt` command provided with an arg", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Rest: []string{"adopt", "foo"},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, nil)
		testing.AssertSpyCommandsCalls(*cmds, &testing.SpyCommandsCallNumber{Adopt: 1})
		testing.AssertSpyDisplaysCalls(*displays, nil)
		Expect(cmds.Calls.Adopt[0].Args).To(Equal([]any{commands.AdoptArgs{From: "foo"}}))
	})

	It("should print help if command is not supported", func() {
		src.App(src.Args{
			CmdArgs: src.CmdArgs{
				Rest: []string{"foo"},
			},
			Displays: displays,
			Commands: cmds,
			Logger:   logger,
		})

		testing.AssertSpyLoggerCalls(*logger, &testing.SpyLoggerCallNumber{Warnnl: 1})
		testing.AssertSpyCommandsCalls(*cmds, nil)
		testing.AssertSpyDisplaysCalls(*displays, &testing.SpyDisplaysCallNumber{Help: 1})
		Expect(logger.Calls.Warnnl[0].Args).To(Equal([]any{"Command %s not found", "foo"}))
	})
})
